### ISSUE TITLE
useless-inner-function: false positive when the inner function is only referenced in the context expression of an as-bound with statement

### ISSUE THREAD

**Issue by roosteer** [returntocorp/sgrep-rules#4059]
### Summary

`python.lang.maintainability.useless-inner-function` reports *"function `inner` is defined inside a function but never used"* for an inner function whose only reference sits in the context expression of a `with` statement **that carries an `as` clause**. Removing the `as` binding makes the false positive disappear; the inner function is referenced in both forms.

### Reproduction

`suspect.py`:

```python
from contextlib import contextmanager


@contextmanager
def ctx(*args):
    yield None


def outer():
    def inner():
        return 1

    with ctx(inner) as cm:      # only reference to `inner`
        print(cm)
```

```
semgrep --config python/lang/maintainability/useless-innerfunction.yaml suspect.py
```

Observed:

```
suspect.py
❯❯❱ python.lang.maintainability.useless-inner-function
       6┆ def outer():
       7┆     def inner():
       8┆         return 1
       9┆
      10┆     with ctx(inner) as cm:      # only reference to `inner`
      11┆         print(cm)
function `inner` is defined inside a function but never used
```

Expected: no finding — the name `inner` is used.

Dropping only the `as cm` clause (keeping `with ctx(inner):`) reports nothing, which locates the trigger in the `with` head of an `as`-bound statement:

```python
def outer():
    def inner():
        return 1
    with ctx(inner):        # no finding, correctly
        pass
```

### Additional data

* Shape confirmed in a real test suite: an inner `cli(cmd, **kwargs)` helper passed as `with patch("...", side_effect=cli) as cli_run:` is reported as never used (3 hits in one file).
* Also fires when the same inner function is defined inside a **method** (class + method nesting does not change it).
* Reference in other compound-statement heads does **not** fire: `if head(...)`, `for head(...)`, `while head(...)`, `try/except` body. A reference in the decorator list of a nested function does fire (separate shape, may or may not be intended).
* The `pattern-not-inside: <... $FF ...>` clause is what should exempt the reference; it appears not to see a reference inside the context expression of an `as`-bound `with`.

### Environment

* semgrep 1.145.0 (same behaviour with the rule file at `develop`)
* rule file: `python/lang/maintainability/useless-innerfunction.yaml`, unmodified from this repository
* Python 3.11
